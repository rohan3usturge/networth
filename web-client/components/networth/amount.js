import { useEffect, useState } from "react";
import { NUMBER_UTILS } from "../utils";

const Amount = ({ defaultValue, currency, disabled, onChange }) => {
  const initialValue = NUMBER_UTILS.convertNumToMoney(defaultValue, currency);
  const [value, setValue] = useState(initialValue);
  const [isError, setIsError] = useState(false);

  useEffect(() => {
    setValue(NUMBER_UTILS.convertNumToMoney(defaultValue, currency));
  }, [defaultValue, currency]);

  const handleBlur = (e) => {
    const currVal = e.target.value;
    const newVal = NUMBER_UTILS.convertNumToMoney(currVal, currency);
    setValue(newVal);
    if (!isError && currVal != defaultValue && onChange) {
      onChange(currVal);
    }
  };

  const handleFocus = (e) => {
    let currVal = e.target.value;
    let newVal = currVal ? NUMBER_UTILS.convertStrToNum(currVal) : "";
    setValue(newVal);
  };

  const handleChange = (e) => {
    let currVal = e.target.value;
    if (!currVal || !currVal.length) {
      currVal = "0.0";
    }
    currVal = currVal.replace(/[^0-9.]+/g, "");
    if (isNaN(currVal)) {
      setIsError(true);
    } else {
      setIsError(false);
    }
    setValue(currVal);
  };

  return (
    <div className="flex-column">
      <input
        disabled={disabled}
        type="text"
        value={value}
        className="form-control money-input"
        onChange={handleChange}
        onBlur={handleBlur}
        onFocus={handleFocus}
      />
      {isError && (
        <div className="text-danger">
          <i className="fas fa-exclamation mr-1"></i>
          <small>Please enter a valid number</small>
        </div>
      )}
    </div>
  );
};

export { Amount };
