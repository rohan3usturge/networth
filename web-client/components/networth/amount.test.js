import { shallow } from "enzyme";
import { Amount } from "./amount";

jest.mock("next/router", () => {
  return {
    useRouter: () => ({ locale: "en-ca" }),
  };
});

describe("Amount", () => {
  it("should render properly and match snapshot", () => {
    const wrapper = shallow(
      <Amount
        defaultValue="3.00"
        currency="CAD"
        disabled={true}
        onChange={() => {}}
        key="1212"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
