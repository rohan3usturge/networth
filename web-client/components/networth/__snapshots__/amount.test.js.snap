// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Amount should render properly and match snapshot 1`] = `
<div
  className="flex-column"
>
  <input
    className="form-control money-input"
    disabled={true}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="text"
    value="CA$3.00"
  />
</div>
`;
